parquet-column/src/main/java/org/apache/parquet/io/RecordReaderImplementation.java

```diff
@@ -424,13 +434,105 @@ public T read() {
     return record;
   }
 
+  private State getState(MessageType schema) {
+    for (State state : states) {
+        if (schema.containsField(state.fieldPath[0])) {
+          return state;
+        }
+    }
+    throw new IllegalArgumentException("Couldn't find state for schema " + schema);
+  }
+
   @Override
-  public boolean shouldSkipCurrentRecord() {
-    return shouldSkipCurrentRecord;
+  public void readVectors(ColumnVector[] vectors, MessageType[] columnSchemas, long current, long total) {
+      for (int i = 0 ; i < columnSchemas.length; i++) {
+        MessageType schema  = columnSchemas[i];
+        State state = getState(schema);
+        readVector(state.column, vectors[i], current, total);
+      }
   }
 
-  private static void log(String string) {
-    LOG.debug(string);
+  @Override
+  public void readVector(ObjectColumnVector<T> vector, long current, long total) {
+    int index = 0;
+    for ( ; index < ColumnVector.DEFAULT_VECTOR_LENGTH; index++, current++) {
+
+      if (current >= total) {
+        ((ColumnVector) vector).setNumberOfValues(index);
+        return;
+      }
+
+      T record = read();
+
+      if (shouldSkipCurrentRecord) {
+        if (DEBUG) {
+          LOG.debug("skipping record");
+        }
+        vector.isNull[index] = true;
+      }
+      vector.values[index] = record;
+    }
+
+    ((ColumnVector) vector).setNumberOfValues(index);
```

Comment:
The problem is that setNumberOfValues is package-private. I actually don't see why it even works to cast like this. If we intend to use this outside of the package, I think it should be public. The only public API part should be the ColumnVector interface, while the abstract base class (that carries this method) should be considered private API. We'll have to separate those two if they are currently just one abstract class.